tests/test_DG_advection.py

```diff
@@ -0,0 +1,107 @@
+                                 refinement_level=refinements)
+    global_normal = Expression(("x[0]/sqrt(x[0]*x[0]+x[1]*x[1]+x[2]*x[2])",
+                                "x[1]/sqrt(x[0]*x[0]+x[1]*x[1]+x[2]*x[2])",
+                                "x[2]/sqrt(x[0]*x[0]+x[1]*x[1]+x[2]*x[2])"))
+    mesh.init_cell_orientations(global_normal)
+
+    fieldlist = ['u','D']
+    timestepping = TimesteppingParameters(dt=dt)
+    output = OutputParameters(dumpfreq=150, dirname='results/tests/'+dirname)
+    parameters = ShallowWaterParameters()
+
+    state = ShallowWaterState(mesh, vertical_degree=None, horizontal_degree=2,
+                              family="BDM",
+                              timestepping=timestepping,
+                              output=output,
+                              parameters=parameters,
+                              fieldlist=fieldlist)
+
+    # interpolate initial conditions
+    u0 = Function(state.V[0], name="velocity")
+    x = SpatialCoordinate(mesh)
+    uexpr = as_vector([-x[1], x[0], 0.0])
+    u0.project(uexpr)
+
+    if vector:
+        VectorDGSpace = VectorFunctionSpace(mesh, "DG", 1)
+        f = Function(VectorDGSpace, name="f")
+        fexpr = Expression(("exp(-pow(x[2],2) - pow(x[1],2))", "0.0", "0.0"))
+        f_end = Function(VectorDGSpace)
+        f_end_expr = Expression(("exp(-pow(x[2],2) - pow(x[0],2))","0","0"))
+    else:
+        f = Function(state.V[1], name='f')
+        fexpr = Expression("exp(-pow(x[2],2) - pow(x[1],2))")
+        f_end = Function(state.V[1])
+        f_end_expr = Expression("exp(-pow(x[2],2) - pow(x[0],2))")
+
+    f.interpolate(fexpr)
+    f_end.interpolate(f_end_expr)
+
+    return state, u0, f, f_end
+
+
+def run(continuity=False, vector=False):
+
+    state, u0, f, f_end = setup_DGadvection(dirname="DGAdvection/continuity" + str(continuity))
+
+    dt = state.timestepping.dt
+    tmax = pi/4.
+    t = 0.
+    f_advection = DGAdvection(state, f.function_space(), continuity=continuity)
+
+    fp1 = Function(f.function_space())
+    f_advection.ubar.assign(u0)
+
+    dumpcount = itertools.count()
+    outfile = File(state.output.dirname+".pvd")
+    outfile.write(f)
+
+    while t < tmax + 0.5*dt:
+        t += dt
+        for i in range(2):
+            f_advection.apply(f, fp1)
+            f.assign(fp1)
+
+        if(next(dumpcount) % state.output.dumpfreq) == 0:
+            outfile.write(f)
+
+    f_err = Function(f.function_space()).assign(f_end - f)
+    return f_err
+
+
+def test_dgadvection_scalar():
+
+    f_err = run()
+    assert(abs(f_err.dat.data.max()) < 1.5e-2)
+
+
+def test_dgadvection_continuity_scalar():
+
+    f_err = run(continuity=True)
+    assert(abs(f_err.dat.data.max()) < 1.5e-2)
+
+
+def test_dgadvection_vector():
+
+    f_err = run(vector=True)
+    assert(abs(f_err.dat.data.max()) < 1.5e-2)
+
+
+def test_dgadvection_continuity_vector():
+
+    f_err = run(continuity=True, vector=True)
+    assert(abs(f_err.dat.data.max()) < 1.5e-2)
```

Comment:
Nice. I had a feeling you'd tell me there was an easier way.


________________________________
From: Lawrence Mitchell <notifications@github.com>
Sent: 12 April 2016 18:05
To: firedrakeproject/dcore
Cc: Shipton, Jemma
Subject: Re: [firedrakeproject/dcore] Dg advection (#23)


In tests/test_DG_advection.py<https://github.com/firedrakeproject/dcore/pull/23#discussion_r59413240>:

> +def test_dgadvection_continuity_scalar():
> +
> +    f_err = run(continuity=True)
> +    assert(abs(f_err.dat.data.max()) < 1.5e-2)
> +
> +
> +def test_dgadvection_vector():
> +
> +    f_err = run(vector=True)
> +    assert(abs(f_err.dat.data.max()) < 1.5e-2)
> +
> +
> +def test_dgadvection_continuity_vector():
> +
> +    f_err = run(continuity=True, vector=True)
> +    assert(abs(f_err.dat.data.max()) < 1.5e-2)


AKA:

import pytest

@pytest.parametrize("vector", [False, True], ids=["scalar", "vector"])
@pytest.parametrize("continuity", [False, True])
def test_dgadvection(vector, continuity):
    f_err = run(continuity=continuity, vector=vector)
    assert abs(f_err.dat.data_ro.max()) < 1.5e-2

-
You are receiving this because you authored the thread.
Reply to this email directly or view it on GitHub<https://github.com/firedrakeproject/dcore/pull/23/files/f322e474cebe58e9debfb0d41281be1f458c7392#r59413240>